pennylane/math/matrix_manipulation.py

```diff
@@ -145,114 +145,131 @@ def expand_matrix(base_matrix, wires, wire_order=None, sparse_format="csr"):
     return mat
 
 
-def reduce_matrices(
-    mats_and_wires_gen: Generator[Tuple[np.ndarray, Wires], None, None], reduce_func: callable
-) -> Tuple[np.ndarray, Wires]:
-    """Apply the given ``reduce_func`` cumulatively to the items of the ``mats_and_wires_gen``
-    generator, from left to right, so as to reduce the sequence to a tuple containing a single
-    matrix and the wires it acts on.
+def _sparse_expand_matrix(base_matrix, wires, wire_order, format="csr"):
+    """Re-express a sparse base matrix acting on a subspace defined by a set of wire labels
+    according to a global wire order.
 
     Args:
-        mats_and_wires_gen (Generator): generator of tuples containing the matrix and the wires of
-            each operator
-        reduce_func (callable): function used to reduce the sequence of operators
+        base_matrix (scipy.sparse.spmatrix): base matrix to expand
+        wires (Iterable): wires determining the subspace that base matrix acts on; a base matrix of
+            dimension :math:`2^n` acts on a subspace of :math:`n` wires
+        wire_order (Iterable): global wire order, which has to contain all wire labels in ``wires``, but can also
+            contain additional labels
+        format (str): string representing the preferred scipy sparse matrix format to cast the expanded
+            matrix too
 
     Returns:
-        Tuple[tensor, Wires]: a tuple containing the reduced matrix and the wires it acts on
+        tensor_like: expanded matrix
     """
+    base_matrix.eliminate_zeros()
 
-    def expand_and_reduce(op1_tuple: Tuple[np.ndarray, Wires], op2_tuple: Tuple[np.ndarray, Wires]):
-        mat1, wires1 = op1_tuple
-        mat2, wires2 = op2_tuple
-        expanded_wires = wires1 + wires2
-        mat1 = qml.math.expand_matrix(mat1, wires1, wire_order=expanded_wires)
-        mat2 = qml.math.expand_matrix(mat2, wires2, wire_order=expanded_wires)
-        return reduce_func(mat1, mat2), expanded_wires
+    wires = wires.tolist() if isinstance(wires, qml.wires.Wires) else list(copy.copy(wires))
 
-    reduced_mat, final_wires = reduce(expand_and_reduce, mats_and_wires_gen)
+    expanded_wires = []
+    mats = []
+    op_wires_in_list = False
```

Comment:
I'm trying to figure out if we can make this chunk a little easier to read.
```suggestion
    base_matrix_encoutered= False
```

It might also be worthwhile leaving some comments, like:
```
# separate the wire order into
# 1) identities before the matrix in wire order
# 2) the matrix
# 3) identities after or in the middle of the matrix in wire order
# Any identity wires in between matrix wires will be added after the matrix in `expanded_wires
# and permutation will bring them inside
```

Looking at this now, we actually only need to apply the permutation matrix to the base matrix and any identities that might be inside the base matrix.

For example, if we had `wires=(4,2)` and `wire_order=(1,2,3,4,5)`, we would only need to apply a permutation matrix for wires 2,3,4.  We could potentially leave kron-ing the pre- and post- identities till after applying the permutation. Think that seems reasonable?

Reply:
Name changed! 👍

Reply:
Good catch! I will do this instead.

Reply:
Permuting the matrix first reduces performance.

Reply:
Added comments.

Reply:
Now the matrix is permuted first